### PR TITLE
BTTRIBEV2-10838-Toast_Message_-_New_Component

### PR DESCRIPTION
**Pull request by IdanBAnyvision**
### Checklist
- [ ] Does this need a new/update storybook example?
- [ ] Does this need new/update tests?

- [ ] run eslint
- [ ] run test
- [ ] run build

If your answer is yes to any of these, please make sure to include it in your PR.

> NOTE: Please submit all PRs to the `development` branch.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -61,45 +61,45 @@
...
+    "rollup-plugin-peer-deps-external": "^2.2.4",
+    "rollup-plugin-postcss": "^3.1.8",
     "rollup-plugin-sass": "^1.2.2",
     "rollup-plugin-url": "^3.0.1",
-    "url-loader": "^4.1.0",
-    "react": "^16.13.1",
-    "react-dom": "^16.13.1"
+    "url-loader": "^4.1.1",
+    "react": "^16.14.0",
+    "react-dom": "^16.14.0"
```
Comment (orelHAnyvision) [inline package.json:102]: why did you upgrade so many libraries?

File: src/components/ToastMessage/ToastMessage.js [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+import { IconButton, Portal, Animations, Button } from '../../index'
+import { ReactComponent as CloseIcon } from '../../assets/svg/Close.svg'
+import styles from './ToastMessage.module.scss'
+import toastMessageTypeMapper from './ToastMessage.utils'
+
+const ToastMessage = ({ message, type, isUndo, undoCallback, closeIcon, isOpen, className, onClose, hideTimeout }) => {
+  const timerHide = React.useRef()
+
+  const setHideTimeout = useCallback(() => {
+    if (hideTimeout === null) {
```
Comment (orelHAnyvision) [inline src/components/ToastMessage/ToastMessage.js:13]:
```suggestion
    if (!hideTimeout) {
```

File: src/components/ToastMessage/ToastMessage.js [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    )
+  }
+
+  const renderUndoButton = () => {
+    return (
+      isUndo &&
+      <Button className={classNames(styles.undoButton)} onClick={undoCallback} variant='ghost'
+              size='small'>undo</Button>
+    )
+  }
```
Comment (orelHAnyvision) [inline src/components/ToastMessage/ToastMessage.js:63]:
```suggestion
  const renderUndoButton = () => 
      isUndo &&
      <Button className={classNames(styles.undoButton)} onClick={undoCallback} variant='ghost'
              size='small'>undo</Button>
```

File: src/components/ToastMessage/ToastMessage.js [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        {closeIcon}
+      </IconButton>
+    )
+  }
+
+  const renderUndoButton = () => {
+    return (
+      isUndo &&
+      <Button className={classNames(styles.undoButton)} onClick={undoCallback} variant='ghost'
+              size='small'>undo</Button>
```
Comment (orelHAnyvision) [inline src/components/ToastMessage/ToastMessage.js:61]: please use language service for texts

File: src/components/ToastMessage/ToastMessage.js [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+      >
+        {closeIcon}
+      </IconButton>
+    )
+  }
+
+  const renderUndoButton = () => {
+    return (
+      isUndo &&
+      <Button className={classNames(styles.undoButton)} onClick={undoCallback} variant='ghost'
```
Comment (orelHAnyvision) [inline src/components/ToastMessage/ToastMessage.js:60]: if only one class you can remove classNames function

File: src/components/ToastMessage/ToastMessage.js [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+              size='small'>undo</Button>
+    )
+  }
+
+  const classes = classNames(styles.toastMessage, className, {
+    [styles[type]]: !!type,
```
Comment (orelHAnyvision) [inline src/components/ToastMessage/ToastMessage.js:66]: what can be 'type' ?

File: src/components/ToastMessage/ToastMessage.js [hunk 7]
```diff
@@ -0,0 +1,128 @@
+import React, { useEffect, useCallback } from 'react'
+import propTypes from 'prop-types'
+import classNames from 'classnames'
+import { IconButton, Portal, Animations, Button } from '../../index'
+import { ReactComponent as CloseIcon } from '../../assets/svg/Close.svg'
+import styles from './ToastMessage.module.scss'
+import toastMessageTypeMapper from './ToastMessage.utils'
```
Comment (orelHAnyvision) [inline src/components/ToastMessage/ToastMessage.js:7]: reorder
3d libraries
functions
internal components
scss

File: src/components/ToastMessage/ToastMessage.js [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+
+const ToastMessage = ({ message, type, isUndo, undoCallback, closeIcon, isOpen, className, onClose, hideTimeout }) => {
+  const timerHide = React.useRef()
```
Comment (orelHAnyvision) [inline src/components/ToastMessage/ToastMessage.js:10]: import useRef

```suggestion
  const timerHide = useRef()
```

File: src/components/ToastMessage/ToastMessage.module.scss [hunk 9]
```diff
@@ -0,0 +1,77 @@
+@import '~@anyvision/style-guide';
+
+.toastMessage {
+  display: flex;
+  @include av-elevation(temporary);
+  min-width: av-size(sz-320);
+  max-width: av-size(sz-480);
+  height: 72px;
```
Comment (orelHAnyvision) [inline src/components/ToastMessage/ToastMessage.module.scss:8]: use var please. if doesn't exists please talk to Michael

File: src/components/ToastMessage/ToastMessage.module.scss [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+  }
+
+  &.info {
+    box-shadow: inset 0 -6px 0 0 av-color(primary, av-opacity(divider));
+    border: solid 1px av-color(primary);
+  }
+
+  &.error {
+    box-shadow: inset 0 -6px 0 0 av-color(error, av-opacity(divider));
+    border: solid 1px av-color(error);
```
Comment (orelHAnyvision) [inline src/components/ToastMessage/ToastMessage.module.scss:33]: use var please. if doesn't exists please talk to Michael

File: src/components/ToastMessage/ToastMessage.module.scss [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+  color: av-color(trueblack);
+
+  .leadingIcon {
+    margin-right: av-space(tiny);
+    display: flex;
+    align-items: center;
+
+    & > svg {
+      height: 2.5em;
+      width: 2.5em;
```
Comment (orelHAnyvision) [inline src/components/ToastMessage/ToastMessage.module.scss:60]: usr vars

File: src/components/ToastMessage/ToastMessage.module.scss [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+  display: flex;
+
+  .closeIcon {
+    color: av-color(trueblack);
+    margin-left: av-space(tiny);
+  }
+
+  .undoButton {
+    font-weight: av-font-weight(bold);
+    margin: 0 10px;
```
Comment (orelHAnyvision) [inline src/components/ToastMessage/ToastMessage.module.scss:75]: var.... av-space

File: src/components/ToastMessage/ToastMessage.stories.js [hunk 13]
```diff
@@ -0,0 +1,98 @@
+import React, { useState, useCallback } from 'react'
+import ToastMessage from './ToastMessage'
+import { ReactComponent as SunIcon } from '../../assets/svg/Sun.svg'
+import { ReactComponent as CheckIcon } from '../../assets/svg/Check.svg'
+import { centerDecorator } from '../../utils/storybook/decorators'
+import styles from '../../storybook/index.module.scss'
+import { Button } from '../Button'
```
Comment (orelHAnyvision) [inline src/components/ToastMessage/ToastMessage.stories.js:7]: reorder

File: src/components/ToastMessage/ToastMessage.stories.js [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+export const WithUndoButton = () => {
+  const [isOpen, setIsOpen] = useState(false)
+  return (
+    <>
+      <Button onClick={() => setIsOpen(true)}>With Undo Button</Button>
+      <ToastMessage
+        isOpen={isOpen}
+        onClose={useCallback(() => setIsOpen(false), [])}
+        message='This Is a Toast Message with undo button'
+        isUndo={true}
```
Comment (orelHAnyvision) [inline src/components/ToastMessage/ToastMessage.stories.js:88]:
```suggestion
        isUndo
```

File: src/components/ToastMessage/ToastMessage.test.js [hunk 15]
```diff
@@ -0,0 +1,106 @@
+import React from 'react'
+import { render, fireEvent, waitFor, screen } from '@testing-library/react'
+import ToastMessage from './ToastMessage'
+import { ReactComponent as SunIcon } from '../../assets/svg/Sun.svg'
```
Comment (orelHAnyvision) [inline src/components/ToastMessage/ToastMessage.test.js:4]: import from anv-icons...

File: src/components/ToastMessage/ToastMessage.js [hunk 16]
```diff
@@ -0,0 +1,128 @@
+import React, { useEffect, useCallback } from 'react'
+import propTypes from 'prop-types'
+import classNames from 'classnames'
+import { IconButton, Portal, Animations, Button } from '../../index'
+import { ReactComponent as CloseIcon } from '../../assets/svg/Close.svg'
```
Comment (orelHAnyvision) [inline src/components/ToastMessage/ToastMessage.js:5]: import from anv-icons

**Review by orelHAnyvision — CHANGES_REQUESTED** [review]
please also add space before/after brackets {}

File: package.json (hunk 1 above)
Comment (IdanBAnyvision) [reply package.json:102]: @orelHAnyvision - I didn't really upgrade any library, I just updated it to be the version that was already installed by npm.

File: src/components/ToastMessage/ToastMessage.js (hunk 6 above)
Comment (IdanBAnyvision) [reply src/components/ToastMessage/ToastMessage.js:66]: @orelHAnyvision - type can be one of these - 'success', 'info', 'error', 'alert'

File: src/components/ToastMessage/ToastMessage.module.scss (hunk 9 above)
Comment (IdanBAnyvision) [reply src/components/ToastMessage/ToastMessage.module.scss:8]: @orelHAnyvision - I talked to Michael and he asked me to add 2 sizes (sz-64 and sz-8 in order to get 72), this is how i did this:
`  height: calc(#{av-size(sz-64)} + #{av-size(sz-8)});`

File: src/components/ToastMessage/ToastMessage.module.scss (hunk 10 above)
Comment (IdanBAnyvision) [reply src/components/ToastMessage/ToastMessage.module.scss:33]: @orelHAnyvision - for what exactly should I use var? 
for the `-6px` ?
If so, this size doesn't exist.
and we don't a way to combine 2 existing sizes to get 6px.

**Comment by IdanBAnyvision** [comment]
@orelHAnyvision - I fixed all your comments. Thanks for the review.

regarding this comment - _add space before/after brackets {}_ - Ira said that this is not needed in this repo...

File: src/components/ToastMessage/ToastMessage.js (hunk 6 above)
Comment (orelHAnyvision) [reply src/components/ToastMessage/ToastMessage.js:64]: so !! will always be true.... whats is the meaning of this condition?
also add proptypes of oneOf for that type

File: src/components/ToastMessage/ToastMessage.stories.js [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+  )
+}
+
+export const WithUndoButton = () => {
+  const [isOpen, setIsOpen] = useState(false)
+  return (
+    <>
+      <Button onClick={() => setIsOpen(true)}>With Undo Button</Button>
+      <ToastMessage
+        isOpen
```
Comment (orelHAnyvision) [inline src/components/ToastMessage/ToastMessage.stories.js:83]: ???? why isOpen always true?
I think you confused..

Comment (IdanBAnyvision) [reply src/components/ToastMessage/ToastMessage.stories.js]: @orelHAnyvision - fixed and pushed a new commit

**Comment by IdanBAnyvision** [comment]
@orelHAnyvision - fixed and pushed a new commit

File: CHANGELOG.md [hunk 18]
```diff
@@ -50,6 +50,13 @@
 - New Dropdown component
 - Added DualThumb for RangeSlider
 - New Tree component
+- Added new `<ToastMessage />` component - `src/components/ToastMessage`
+
+### packages updates:
+- storybook 6.1.14 -> 6.1.21
```
Comment (IrenaSAnyVision) [inline CHANGELOG.md:56]: I think there is no need to put it there

File: package.json (hunk 1 above)
Comment (orelHAnyvision) [reply package.json:103]: do you use node 12 or 14?

File: src/components/ToastMessage/ToastMessage.module.scss [hunk 19]
```diff
@@ -0,0 +1,83 @@
+@import '~@anyvision/style-guide';
+
+:root {
+  --box-shadow-vertical-length: -6px;
+  --svg-size: 2.5em;
+  --undo-button-margin: 10px;
```
Comment (orelHAnyvision) [inline src/components/ToastMessage/ToastMessage.module.scss:6]: av space

File: src/components/ToastMessage/ToastMessage.module.scss [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+  color: av-color(trueblack);
+
+  .leadingIcon {
+    margin-right: av-space(tiny);
+    display: flex;
+    align-items: center;
+
+    & > svg {
+      height: var(--svg-size);
+      width: var(--svg-size);
```
Comment (orelHAnyvision) [inline src/components/ToastMessage/ToastMessage.module.scss:66]: av-size...

File: src/components/ToastMessage/ToastMessage.utils.js [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+  },
+  alert: {
+    fillColor: styles.alertFillColor,
+    icon: <ErrorCircleOutlined />,
+  },
+}
+
+const toastMessageTypeMapper = type => {
+  return map[type]
+}
```
Comment (orelHAnyvision) [inline src/components/ToastMessage/ToastMessage.utils.js:31]:
```suggestion
const toastMessageTypeMapper = type => map[type]
```

File: src/components/ToastMessage/ToastMessage.js [hunk 22]
```diff
@@ -0,0 +1,126 @@
...
+    if (isOpen) {
+      setHideTimeout()
+    }
+
+    return () => {
+      clearTimeout(timerHide.current)
+    }
+  }, [isOpen, hideTimeout, setHideTimeout])
+
+
```
Comment (orelHAnyvision) [inline src/components/ToastMessage/ToastMessage.js:34]:
```suggestion
```

File: CHANGELOG.md (hunk 18 above)
Comment (IdanBAnyvision) [reply CHANGELOG.md:56]: @IrenaSAnyVision - ok - removed.

File: src/components/ToastMessage/ToastMessage.module.scss (hunk 19 above)
Comment (IdanBAnyvision) [reply src/components/ToastMessage/ToastMessage.module.scss:6]: @orelHAnyvision - we don't have these sizes in the styleguide. so I cannot use av-space...

File: src/components/ToastMessage/ToastMessage.module.scss (hunk 20 above)
Comment (IdanBAnyvision) [reply src/components/ToastMessage/ToastMessage.module.scss:66]: @orelHAnyvision - same as my comment above.

**Comment by IdanBAnyvision** [comment]
@orelHAnyvision - ok, I updated a few more things based on Michael feedback, but 1 variable was still needed since this is a specific value for this component:

```
:root {
  --box-shadow-vertical-length: -6px;
}
```

...

File: src/components/ToastMessage/ToastMessage.js (hunk 6 above)
Comment (IdanBAnyvision) [reply src/components/ToastMessage/ToastMessage.js]: @orelHAnyvision - `type: propTypes.oneOf(['success', 'info', 'error', 'alert']),` is already defined as PropTypes.

**Review by orelHAnyvision — APPROVED** [review]

**Pull request merged**